### ISSUE TITLE
Cannot upload a file as a proxy user

### ISSUE THREAD

**Issue by njaffer** [projecthydra/sufia#1551]
When I try to upload a file as a proxy, there is an error: NoMethodError - undefined method `on_behalf_of=' for #<FileSet:0x007fecb1d967f0>

**Comment by njaffer** [comment]
The on_behalf_of method is missing here: https://github.com/projecthydra/sufia/blob/master/app/controllers/concerns/sufia/file_sets_controller_behavior.rb#L76

In sufia 6, the proxy_deposit is under GenericFile Module
https://github.com/projecthydra/sufia/blob/6.x-stable/sufia-models/app/models/concerns/sufia/generic_file/proxy_deposit.rb#L2

wondering should it be under Works folder in Sufia 7, and the name space should be Sufia::Works::ProxyDeposit

Any thoughts? @mjgiarlo , @jcoyne , @hectorcorrea

**Comment by njaffer** [comment]
Found out that the property 'on_behalf_of' is missing in Curation Concerns models. If I add the on_behalf_of property in basic_metadata.rb, it works. Wondering if that is the right place to add it or should it go somewhere else? @grosscol , any ideas?

**Comment by grosscol** [comment]
Not sure.  Is proxy deposit a feature of CurationConcerns? or is it something that got caught up in the porting process and should be a feature of Sufia?

I don't know if there is a better predicate for on_behalf_of, but it looks like basic metadata would be the place for it.  That gets included in both Works and FileSets.  Alternatively, it could be added to an additional basic metadata module in Sufia.

**Comment by njaffer** [comment]
Yeah I added it in file set behavior. https://github.com/projecthydra-labs/curation_concerns/pull/600/files

**Comment by mjgiarlo** [comment]
I don't believe proxy deposit is a feature that should be in CC. It's a Sufia feature. `on_behalf_of` is defined in Sufia here: https://github.com/projecthydra/sufia/blob/master/app/models/concerns/sufia/proxy_deposit.rb It's mixed into Sufia's WorkBehavior here: https://github.com/projecthydra/sufia/blob/c92ff1750f4472faefb4b73270c29321c33f775f/app/models/concerns/sufia/work_behavior.rb#L4 And it gets mixed into GenericWork during generation here: https://github.com/projecthydra/sufia/blob/d515cd314a2d5e9e8179302e69824299e83e43b2/lib/generators/sufia/install_generator.rb#L93

Long story short: proxy deposit is now about works rather than file sets, so it should move out of the FileSet-oriented views and controllers -- you can see all the files that touch upon this using e.g. `ag on_behalf_of` -- and into work-oriented ones. Let me know if you have any questions!

**Comment by njaffer** [comment]
@mjgiarlo: So on_behalf_of should not be here for fileset? https://github.com/projecthydra/sufia/blob/master/app/controllers/concerns/sufia/file_sets_controller_behavior.rb#L76

rather it should be in works?

**Comment by mjgiarlo** [comment]
@njaffer You got it!

**Comment by njaffer** [comment]
@mjgiarlo , what about the user can deposit a file for a work as a proxy user. So should the proxy user be on file_set and work?

**Comment by grosscol** [comment]
Looks like uploading files into an existing work on behalf of another user is not functionality that will be in stock Sufia.  I'm okay with that.

**Comment by njaffer** [comment]
As of now, the UI allows the user to upload a work or upload a file in a work as a proxy user. The confusing part is that if the proxy deposit is moved to works (which is not working) then what about the proxy deposit for files? If the proxy deposit (on_behalf_of) method stays in file_set then both works. Any ideas? @mjgiarlo, @grosscol, @jcoyne, @cam156, @awead

**Comment by mjgiarlo** [comment]
@njaffer @cam156 @grosscol 

Was there discussion elsewhere of @njaffer's questions above? I don't see her questions answered but I see the PR has been merged.

As it stands, what once was a bug looks like it resulted in a feature change. I suspect we have not fully considered the changes required (including UI changes) to build out transfer/proxies for both GenericWorks and FileSets, which is why I was advocating for having that feature available only for GenericWorks in 7.0.0.

If the discussion did happen elsewhere, it'd be great to see that documented here. Thanks, all.

**Comment by cam156** [comment]
@mjgiarlo adding transfer at the work level would be a new feature.  Having still exist at the FileSet level (aka on the upload page) is a bug fix.  The concept of proxy deposit at the work level is a new concept and there is currently no ui or code to back that.

We did have a discussion over IRC yesterday about adding the concept back so we did not loose existing functionality in sufia.  Aka the feature already exists on the upload page in SUfia 6.0, which I believe equates to a FileSet in Sufia 7.0.  For this reason it seemed logical to me to merge the commit.

**Comment by mjgiarlo** [comment]
@cam156 OK, glad to hear it was discussed in IRC. Let's stick with the merge. (I like to see PRs getting merged!)

I suspect we'll have to come back to this before 7.0.0 is released to see how this all works together, given Sufia 7's strong orientation around GenericWorks (vs. FileSets). For example, we now (or will soon) have two upload paths: one that is metadata-first (new, largely from CurationConcerns) and one that is files-first (legacy, largely from Sufia 6), and both create FileSets within GenericWorks. If I use the latter workflow to upload 15 files and expect them to be proxied over to another user, does it make sense for only the FileSets to be proxied (without the containing GenericWork context, which has all the metadata I entered)?

This is why my assumption was that GenericWorks should be the subject of proxies/transfers, not FileSets. I didn't see this as a new feature, necessarily, because Sufia 7's domain model is just different from Sufia 6's -- I see this as providing continuity with Sufia 6 and aligning well with the [migration path](https://github.com/projecthydra/sufia#migrating-data-to-pcdm-in-sufia-7) that's so far been discussed.

**Comment by njaffer** [comment]
@mjgiarlo, like @cam156 said, I had the discussion with her and it made sense to keep the functionality of 'a user can deposit a file on behalf of some other user'. This issue was about the bug-getting error while uploading a file as a proxy user. 

Probably we can open another issue for adding the proxy user for GenericWorks?

**Comment by mjgiarlo** [comment]
@njaffer Yeah, I suggest we may want to wait on the Sufia UI WG's issues, and in implementing those test the UI to make sure proxies and transfers work as expected. Cc: @mtribone @elrayle @hackmastera etc.
